components/script/dom/transformstream.rs

```diff
@@ -997,3 +1003,103 @@ impl TransformStreamMethods<crate::DomTypeHolder> for TransformStream {
+impl Transferable for TransformStream {
+    type Index = MessagePortIndex;
+    type Data = MessagePortImpl;
+
+    fn transfer(&self) -> Result<(MessagePortId, MessagePortImpl), ()> {
+        let global = self.global();
+        let realm = enter_realm(&*global);
+        let comp = InRealm::Entered(&realm);
+        let cx = GlobalScope::get_cx();
+        let can_gc = CanGc::note();
+
+        // Let readable be value.[[readable]].
+        let readable = self.get_readable();
+
+        // Let writable be value.[[writable]].
+        let writable = self.get_writable();
+
+        // If ! IsReadableStreamLocked(readable) is true, throw a "DataCloneError" DOMException.
+        if readable.is_locked() {
+            return Err(());
+        }
+
+        // If ! IsWritableStreamLocked(writable) is true, throw a "DataCloneError" DOMException.
+        if writable.is_locked() {
+            return Err(());
+        }
+
+        // Create the shared port pair
+        let port_1 = MessagePort::new(&global, can_gc);
+        global.track_message_port(&port_1, None);
+        let port_2 = MessagePort::new(&global, can_gc);
+        global.track_message_port(&port_2, None);
+        global.entangle_ports(*port_1.message_port_id(), *port_2.message_port_id());
+
+        // Create a proxy WritableStream wired to port_1
+        let proxy_writable = WritableStream::new_with_proto(&global, None, can_gc);
+        proxy_writable.setup_cross_realm_transform_writable(cx, &port_1, can_gc);
+
+        // Pipe readable into the proxy writable (→ port_1)
+        let pipe1 = readable.pipe_to(
+            cx,
+            &global,
+            &proxy_writable,
+            false,
+            false,
+            false,
+            comp,
+            can_gc,
+        );
+        pipe1.set_promise_is_handled();
+
+        // Create a proxy ReadableStream wired to port_1
+        let proxy_readable = ReadableStream::new_with_proto(&global, None, can_gc);
+        proxy_readable.setup_cross_realm_transform_readable(cx, &port_1, can_gc);
+
+        // Pipe proxy readable (← port_1) into writable
+        let pipe2 =
+            proxy_readable.pipe_to(cx, &global, &writable, false, false, false, comp, can_gc);
+        pipe2.set_promise_is_handled();
+
+        // Set dataHolder.[[readable]] to ! StructuredSerializeWithTransfer(readable, « readable »).
+        // Set dataHolder.[[writable]] to ! StructuredSerializeWithTransfer(writable, « writable »).
+        port_2.transfer()
+    }
+
+    fn transfer_receive(
+        owner: &GlobalScope,
+        id: MessagePortId,
+        port_impl: MessagePortImpl,
+    ) -> Result<DomRoot<Self>, ()> {
+        let can_gc = CanGc::note();
+
+        // Let readableRecord be ! StructuredDeserializeWithTransfer(dataHolder.[[readable]], the current Realm).
+        // Set value.[[readable]] to readableRecord.[[Deserialized]].
+        let readable = ReadableStream::transfer_receive(owner, id, port_impl.clone())?;
```

Comment:
The cloning of the `port_impl` is unnecessary. Instead, you should unwrap the transfer operations here, like is done for `transfer`, and:
- Only `transfer_receive` a single port, 
- `setup_cross_realm_transform_readable` and `setup_cross_realm_transform_writable` with that same port. 

I'm looking into that timeout of "piping through transferred transforms should work", and so far I don't know what caused it, but the cloning of the `port_impl` strikes me as an interesting suspect.